### PR TITLE
Improve caching in ODSP driver

### LINKED ISSUE #4037: Optimize caching strategy in ODSP driver
[Teams chat](https://teams.microsoft.com/l/message/19:9034865f2a4f49b5ab8df4569dbb0e73@thread.skype/1603302504888?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=422665a0-7ad3-4d0d-9171-e8881d0397d9&parentMessageId=1603302504888&teamName=FFX%20(Fluid%20Framework%20and%20Experiences)&channelName=LiveSite%20%E2%9A%A0&createdTime=1603302504888)
Related bug opened on same feedback: https://github.com/microsoft/FluidFramework/issues/4034

OdspDocumentStorageService.initBlobsCache caches all the blobs coming from snapshot.
There are certain problems around existing behavior that needs to be addressed:
1. There is no cache eviction strategy - we keep all the blobs forever. This is not good, we need something here, even rudimentary is fine (like get rid of all entries 10 seconds after loading snapshot).
2. Cache is not shared between multiple container instances using same driver factory. Main client & summarizer double amount of memory used
3. OdspDocumentStorageService.read does not add to cache. That's fine in isolation, but not good for sharing between main client and summarizer, especially given we are moving to snapshots not having all the blobs, so both clients are likely to download more blobs.

I'd break this item into smaller ones, and see if we can get some improvement sooner than later, given the pain it causes.
Even better would be to have caching eviction being governed by memory pressure on the machine, but I'm not sure if it's possible.

Same things apply to trees caching.
Less so to ops - I'd not touch ops at all.

### PR DESCRIPTION
**Pull request by ChumpChief**
Fixes #4037 

In particular, adds to the cache on calls to `read()` and clears the cache after 10 seconds without attempted reads or writes as a basic eviction strategy.  Does not attempt to share cache between separate instances of the container, though I expect this duplication should be less consequential if we're clearing the cache regularly.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 1]
```diff
@@ -237,10 +242,17 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
                     async () => this.epochTracker.fetchAndParseAsJSON<IBlob>(url, { headers }),
                 );
             });
-            blob = response.content;
+            // Verify that we haven't populated the cache in the meantime from a simultaneous read
+            blob = this.blobCache.get(blobid);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:246]: Unless you plan to move this logic to factory scope (i.e. increase scope), I do not think caching these blobs makes sense.
We never read them again

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 2]
```diff
@@ -760,6 +772,18 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
...
     private initBlobsCache(blobs: IBlob[]) {
         blobs.forEach((blob) => this.blobCache.set(blob.id, blob));
+        this.scheduleClearBlobsCache();
+    }
+
+    /**
+     * Stop the current timer for clearing the blob cache (if any) and schedule a new one
+     */
+    private scheduleClearBlobsCache() {
+        const blobCacheTimeoutDuration = 10000;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:782]: This looks fine for starters, but something to keep in mind -continuously setting & resetting timer might be expensive.
In similar places in Office we developed a pattern of building small virtualization layer on top of "hardware" timer where we would not move "hardware" timer if it is too far out, thus amortizing all set/reset calls into O(1).

(I can tell a story what kind of bugs we hit whiteout it, finding some interesting scalability bugs in Windows).

**Review by vladsud — APPROVED** [review]
Looks fine, but I'd love to see removal from cache blobs after reading, not adding them :)
Basically ideal behavior - we get rid of caches as we consume blobs.
It assumes certain behavior from runtime, of course, so there is some risk here.

Also can we also clear tree cache on that timer as well?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:246]: Sounds good, will remove in next iteration.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts (hunk 2 above)
Comment (ChumpChief) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:782]: I'm suspecting we would bottleneck in the blob-reading code before it bottlenecks on the timer set/clear.  But if it comes up we can deal with it then.

**Comment by ChumpChief** [comment]
> Looks fine, but I'd love to see removal from cache blobs after reading, not adding them :)
> Basically ideal behavior - we get rid of caches as we consume blobs.
> It assumes certain behavior from runtime, of course, so there is some risk here.
> 
> Also can we also clear tree cache on that timer as well?

Issue sounded like we needed a quick improvement to reduce memory usage to address issues in the wild so I went with the easy option.  I haven't looked into the storage side of things as much but it sounds like we have semi-conflicting requirements here that need reconciliation (delete-after-read vs. share-amongst-clients).  I'm currently prioritizing delta connection and delta storage over storage but will def look into improvements here when I get into storage investigations.

**Comment by vladsud** [comment]
> Issue sounded like we needed a quick improvement to reduce memory usage to address issues in the wild so I went with the easy option. I haven't looked into the storage side of things as much but it sounds like we have semi-conflicting requirements here that need reconciliation (delete-after-read vs. share-amongst-clients). I'm currently prioritizing delta connection and delta storage over storage but will def look into improvements here when I get into storage investigations.

I do not think requirements are conflicting.
In existing (non sharing) implementation, delete-after-read is the right strategy. share-amongst-clients strategy would benefit from caching after read.

**Pull request merged**